File: server/src/detail.js
const db = require("./db");
const { getSettings, fullNames } = require("./settings");

const nameByIdStmt = db.prepare("SELECT id, name, origin, meaning, rank FROM names WHERE id = ?");
const nameStateStmt = db.prepare("SELECT rounds_survived, eliminated_in FROM name_state WHERE name_id = ?");
const userLabelStmt = db.prepare("SELECT name FROM users WHERE id = ?");
// Only sealed rounds count -- blind reveal means an in-progress round can't
// leak into these numbers.
const keepCountStmt = db.prepare(`
  SELECT COUNT(*) AS n
  FROM set_results sr
  JOIN rounds r ON r.id = sr.round_id
  WHERE sr.name_id = ? AND sr.user_id = ? AND sr.kept = 1 AND r.sealed_at IS NOT NULL
`);
const firstSeenStmt = db.prepare(`
  SELECT MIN(r.number) AS round
  FROM set_results sr
  JOIN rounds r ON r.id = sr.round_id
  WHERE sr.name_id = ? AND r.sealed_at IS NOT NULL
`);

const otherUserId = (id) => (id === 1 ? 2 : 1);

/** Backing data for the name detail sheet (build spec §6 "Shortlist", Change Order 2 §1). */
function getNameDetail(nameId, viewerId) {
  const n = nameByIdStmt.get(nameId);
  if (!n) return null;

  const state = nameStateStmt.get(nameId);
  const partnerId = otherUserId(viewerId);
  const partner = userLabelStmt.get(partnerId);
  const firstSeen = firstSeenStmt.get(nameId);

  return {
    id: n.id,
    name: n.name,
    origin: n.origin,
    meaning: n.meaning,
    auRank: n.rank,
    fullNames: fullNames(n.name, getSettings()),
    roundsSurvived: state?.rounds_survived ?? 0,
    eliminatedIn: state?.eliminated_in ?? null,
    youKeptCount: keepCountStmt.get(nameId, viewerId).n,
    partnerLabel: partner.name,
    partnerKeptCount: keepCountStmt.get(nameId, partnerId).n,
    firstSeenRound: firstSeen?.round ?? null,
  };
}

module.exports = { getNameDetail };

File: server/detail.py
"""Ported from server/src/detail.js. Backing data for the name detail sheet
(build spec §6 "Shortlist", Change Order 2 §1).
"""

from settings import full_names, get_settings

NAME_BY_ID_SQL = "SELECT id, name, origin, meaning, rank FROM names WHERE id = ?"
NAME_STATE_SQL = "SELECT rounds_survived, eliminated_in FROM name_state WHERE name_id = ?"
USER_LABEL_SQL = "SELECT name FROM users WHERE id = ?"
# Only sealed rounds count -- blind reveal means an in-progress round can't
# leak into these numbers.
KEEP_COUNT_SQL = """
    SELECT COUNT(*) AS n
    FROM set_results sr
    JOIN rounds r ON r.id = sr.round_id
    WHERE sr.name_id = ? AND sr.user_id = ? AND sr.kept = 1 AND r.sealed_at IS NOT NULL
"""
FIRST_SEEN_SQL = """
    SELECT MIN(r.number) AS round
    FROM set_results sr
    JOIN rounds r ON r.id = sr.round_id
    WHERE sr.name_id = ? AND r.sealed_at IS NOT NULL
"""


def _other_user_id(user_id):
    return 2 if user_id == 1 else 1


def get_name_detail(conn, name_id, viewer_id):
    n = conn.execute(NAME_BY_ID_SQL, (name_id,)).fetchone()
    if not n:
        return None

    state = conn.execute(NAME_STATE_SQL, (name_id,)).fetchone()
    partner_id = _other_user_id(viewer_id)
    partner = conn.execute(USER_LABEL_SQL, (partner_id,)).fetchone()
    first_seen = conn.execute(FIRST_SEEN_SQL, (name_id,)).fetchone()

    return {
        "id": n["id"],
        "name": n["name"],
        "origin": n["origin"],
        "meaning": n["meaning"],
        "auRank": n["rank"],
        "fullNames": full_names(n["name"], get_settings(conn)),
        "roundsSurvived": state["rounds_survived"] if state else 0,
        "eliminatedIn": state["eliminated_in"] if state else None,
        "youKeptCount": conn.execute(KEEP_COUNT_SQL, (name_id, viewer_id)).fetchone()["n"],
        "partnerLabel": partner["name"],
        "partnerKeptCount": conn.execute(KEEP_COUNT_SQL, (name_id, partner_id)).fetchone()["n"],
        "firstSeenRound": first_seen["round"] if first_seen else None,
    }
